File: sdk/npm/src/providers/tempgmailer.ts
/**
 * tempgmailer.com 临时邮箱渠道
 *
 * 流程：
 *   1. GET 首页获取 CSRF token 和 Laravel session cookie
 *   2. POST /get-gmail 创建 Gmail 临时邮箱
 *   3. POST /get-inbox 获取邮件列表
 */
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tempgmailer";
const BASE_URL = "https://tempgmailer.com";

const COMMON_HEADERS: Record<string, string> = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36",
  Accept: "application/json, text/plain, */*",
  "Content-Type": "application/json",
  "X-Requested-With": "XMLHttpRequest",
  "X-TempGmailer-Auth": "frontend",
  Origin: BASE_URL,
  Referer: `${BASE_URL}/`,
};

/** 会话信息：CSRF token + session cookie */
interface TempGmailerSession {
  csrfToken: string;
  cookie: string;
}

/** /get-gmail 响应 */
interface CreateResponse {
  success?: boolean;
  data?: {
    email?: string;
  };
}

/** /get-inbox 中的邮件项 */
interface InboxMessage {
  from?: { address?: string; name?: string } | string;
  subject?: string;
  intro?: string;
  body?: string;
  createdAt?: string;
}

/** /get-inbox 响应 */
interface InboxResponse {
  success?: boolean;
  data?: {
    email?: string;
    messages?: InboxMessage[];
  };
}

/**
 * 从 Set-Cookie 头中提取 cookie 字符串
 */
function cookieHeader(headers: Headers): string {
  const getSetCookie = (headers as any).getSetCookie?.bind(headers);
  const values: string[] =
    typeof getSetCookie === "function"
      ? getSetCookie()
      : headers.get("set-cookie")
        ? [headers.get("set-cookie") as string]
        : [];
  return values
    .map((value) => value.split(";")[0])
    .filter(Boolean)
    .join("; ");
}

/**
 * 从 HTML 中提取 CSRF token
 * 匹配 <meta name="csrf-token" content="...">
 */
function extractCsrfToken(html: string): string {
  const match = html.match(
    /<meta\s+name="csrf-token"\s+content="([^"]+)"/,
  );
  return match ? match[1] : "";
}

/**
 * 初始化会话：获取 CSRF token 和 session cookie
 */
async function initSession(): Promise<TempGmailerSession> {
  const response = await fetchWithTimeout(BASE_URL, {
    headers: {
      Accept:
        "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      "User-Agent": COMMON_HEADERS["User-Agent"],
    },
  });
  if (!response.ok) {
    throw new Error(`tempgmailer home: ${response.status}`);
  }

  const cookie = cookieHeader(response.headers);
  const html = await response.text();
  const csrfToken = extractCsrfToken(html);

  if (!cookie || !csrfToken) {
    throw new Error("tempgmailer: 获取 CSRF token 或 session cookie 失败");
  }

  return { csrfToken, cookie };
}

/**
 * 编码会话信息为 token 字符串
 */
function encodeSession(session: TempGmailerSession): string {
  return JSON.stringify(session);
}

/**
 * 从 token 字符串解码会话信息
 */
function decodeSession(token: string): TempGmailerSession {
  const data = JSON.parse(token) as Partial<TempGmailerSession>;
  if (!data.cookie || !data.csrfToken) {
    throw new Error("tempgmailer: 无效的 session token");
  }
  return { cookie: data.cookie, csrfToken: data.csrfToken };
}

/**
 * 创建 tempgmailer 临时邮箱
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const session = await initSession();

  const response = await fetchWithTimeout(`${BASE_URL}/get-gmail`, {
    method: "POST",
    headers: {
      ...COMMON_HEADERS,
      "X-CSRF-TOKEN": session.csrfToken,
      Cookie: session.cookie,
    },
    body: JSON.stringify({ refresh: true, adblock: 0 }),
  });

  if (!response.ok) {
    throw new Error(`tempgmailer get-gmail: ${response.status}`);
  }

  const data = (await response.json()) as CreateResponse;
  if (!data.success || !data.data?.email) {
    throw new Error("tempgmailer: 创建邮箱失败");
  }

  return {
    channel: CHANNEL,
    email: data.data.email,
    token: encodeSession(session),
  };
}

/**
 * 获取 tempgmailer 邮件列表
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const session = decodeSession(token);

  const response = await fetchWithTimeout(`${BASE_URL}/get-inbox`, {
    method: "POST",
    headers: {
      ...COMMON_HEADERS,
      "X-CSRF-TOKEN": session.csrfToken,
      Cookie: session.cookie,
    },
    body: JSON.stringify({ email, adblock: 0 }),
  });

  if (!response.ok) {
    throw new Error(`tempgmailer get-inbox: ${response.status}`);
  }

  const data = (await response.json()) as InboxResponse;
  if (!data.success) {
    throw new Error("tempgmailer: 获取收件箱失败");
  }

  const messages = data.data?.messages ?? [];
  const emails: Email[] = [];

  for (const msg of messages) {
    /* 解析发件人地址 */
    let fromAddr = "";
    if (typeof msg.from === "string") {
      fromAddr = msg.from;
    } else if (msg.from?.address) {
      fromAddr = msg.from.address;
    }

    /* body 为完整 HTML，intro 为纯文本摘要 */
    const htmlContent = msg.body ?? msg.intro ?? "";
    const textContent = msg.intro ?? "";

    emails.push(
      normalizeEmail(
        {
          from: fromAddr,
          to: email,
          subject: msg.subject ?? "",
          html: htmlContent,
          text: textContent,
          date: msg.createdAt ?? "",
        },
        email,
      ),
    );
  }

  return emails;
}

File: sdk/rust/src/providers/tempgmailer.rs
/*!
 * TempGmailer 渠道实现
 * 网站: https://tempgmailer.com
 * 提供真实 @gmail.com 地址（dot trick）
 * 需要 CSRF token + Laravel session cookie 认证
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::Value;

const BASE_URL: &str = "https://tempgmailer.com";

fn browser_ua() -> &'static str {
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36"
}

/// 从首页 HTML 中提取 CSRF token 和 session cookie
fn init_session() -> Result<(String, String), String> {
    block_on(async {
        let resp = http_client()
            .get(BASE_URL)
            .header(
                "Accept",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            )
            .header("User-Agent", browser_ua())
            .send()
            .await
            .map_err(|e| format!("tempgmailer init session failed: {}", e))?;

        // 提取 Set-Cookie 中的 session cookie
        let mut cookie_parts: Vec<String> = Vec::new();
        for value in resp.headers().get_all("set-cookie") {
            let cookie_str = value.to_str().unwrap_or("");
            if let Some(name_value) = cookie_str.split(';').next() {
                cookie_parts.push(name_value.to_string());
            }
        }

        let cookies = cookie_parts.join("; ");
        if cookies.is_empty() {
            return Err("tempgmailer: failed to get session cookie".into());
        }

        // 从 HTML 中提取 CSRF token
        let html = resp
            .text()
            .await
            .map_err(|e| format!("tempgmailer read html failed: {}", e))?;

        let csrf_token = extract_csrf_token(&html)
            .ok_or_else(|| "tempgmailer: failed to extract csrf-token from HTML".to_string())?;

        Ok((csrf_token, cookies))
    })
}

/// 从 HTML meta 标签中提取 CSRF token
fn extract_csrf_token(html: &str) -> Option<String> {
    // 匹配 <meta name="csrf-token" content="...">
    let marker = "name=\"csrf-token\"";
    let pos = html.find(marker)?;
    let after = &html[pos..];
    let content_start = after.find("content=\"")? + "content=\"".len();
    let content_slice = &after[content_start..];
    let end = content_slice.find('"')?;
    Some(content_slice[..end].to_string())
}

/// 发送带认证头的 POST 请求
fn post_json(csrf_token: &str, cookies: &str, path: &str, body: Value) -> Result<String, String> {
    let csrf_token = csrf_token.to_string();
    let cookies = cookies.to_string();
    let path = path.to_string();
    block_on(async move {
        let resp = http_client()
            .post(format!("{}{}", BASE_URL, path))
            .header("User-Agent", browser_ua())
            .header("Content-Type", "application/json")
            .header("Accept", "application/json, text/plain, */*")
            .header("Origin", BASE_URL)
            .header("Referer", format!("{}/", BASE_URL))
            .header("X-Requested-With", "XMLHttpRequest")
            .header("X-CSRF-TOKEN", &csrf_token)
            .header("X-TempGmailer-Auth", "frontend")
            .header("Cookie", &cookies)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("tempgmailer {} failed: {}", path, e))?;

        let status = resp.status();
        let text = resp
            .text()
            .await
            .map_err(|e| format!("tempgmailer {} read failed: {}", path, e))?;
        if !status.is_success() {
            return Err(format!("tempgmailer {} failed: {} {}", path, status, text));
        }
        Ok(text)
    })
}

/// 创建临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    let (csrf_token, cookies) = init_session()?;
    let text = post_json(
        &csrf_token,
        &cookies,
        "/get-gmail",
        serde_json::json!({ "refresh": true, "adblock": 0 }),
    )?;
    let data: Value = serde_json::from_str(&text).map_err(|e| format!("parse failed: {}", e))?;

    if !data["success"].as_bool().unwrap_or(false) {
        return Err(format!("tempgmailer: generate failed: {}", text));
    }

    let email = data["data"]["email"]
        .as_str()
        .ok_or("tempgmailer: empty email in response")?
        .to_string();

    // 将 CSRF token 和 cookie 组合为 JSON 字符串作为 token
    let token_payload = serde_json::json!({
        "csrfToken": csrf_token,
        "cookies": cookies,
    })
    .to_string();

    Ok(EmailInfo {
        channel: Channel::Tempgmailer,
        email,
        token: Some(token_payload),
        expires_at: None,
        created_at: None,
    })
}

/// 获取邮件列表
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let session: Value =
        serde_json::from_str(token).map_err(|e| format!("parse session token failed: {}", e))?;

    let csrf_token = session["csrfToken"].as_str().unwrap_or("");
    let cookies = session["cookies"].as_str().unwrap_or("");

    let text = post_json(
        csrf_token,
        cookies,
        "/get-inbox",
        serde_json::json!({ "email": email, "adblock": 0 }),
    )?;
    let data: Value = serde_json::from_str(&text).map_err(|e| format!("parse failed: {}", e))?;

    if !data["success"].as_bool().unwrap_or(false) {
        return Err(format!("tempgmailer: get inbox failed: {}", text));
    }

    let messages = data["data"]["messages"].as_array().cloned().unwrap_or_default();

    let mut result = Vec::new();
    for msg in &messages {
        let from_addr = msg["from"]["address"].as_str().unwrap_or("");
        let subject = msg["subject"].as_str().unwrap_or("");
        let html_content = msg["body"].as_str().unwrap_or(msg["intro"].as_str().unwrap_or(""));
        let text_content = msg["intro"].as_str().unwrap_or("");
        let created_at = msg["createdAt"].as_str().unwrap_or("");

        let raw = serde_json::json!({
            "from": from_addr,
            "to": email,
            "subject": subject,
            "text": text_content,
            "html": html_content,
            "date": created_at,
            "isRead": false,
            "attachments": [],
        });
        result.push(normalize_email(&raw, email));
    }

    Ok(result)
}
